json_codec.go

```diff
@@ -0,0 +1,846 @@
+	fieldCodecMap := make(map[string]*codec)
+	for idx, field := range recordTemplate.Fields {
+		var err error
+		fieldCodecs[idx], err = st.buildCodec(recordTemplate.n.namespace(), field.schema)
+		if err != nil {
+			return nil, newCodecBuildError(friendlyName, "record field ought to be codec: %+v", st, err)
+		}
+		fieldCodecMap[field.Name] = fieldCodecs[idx]
+	}
+
+	friendlyName = fmt.Sprintf("record (%s)", recordTemplate.Name)
+
+	c := &codec{
+		nm: recordTemplate.n,
+		df: func(r io.Reader) (interface{}, error) {
+			// Record is Avro JSON encoded as a map with field names as key field values
+			// recursively Avro JSON encoded.
+			// 1. Unmarshal the bytes as regular JSON.
+			// 2. Go through each field and convert from regular JSON to Avro JSON.
+
+			someRecord, _ := NewRecord(recordSchemaRaw(schema), RecordEnclosingNamespace(enclosingNamespace))
+
+			// 1. Unmarshal the bytes as regular JSON.
+			datum, err := jsonDecode(r, friendlyName)
+			if err != nil {
+				return nil, newDecoderError(friendlyName, err)
+			}
+			jsonMap, ok := datum.(map[string]interface{})
+			if !ok {
+				return nil, newCodecBuildError(friendlyName, "Expected JSON map but got %T", datum)
+			}
+
+			// 2. Go through each field and convert from regular JSON to Avro JSON.
+			for key, value := range jsonMap {
+				b, err := json.Marshal(value)
+				if err != nil {
+					return nil, newDecoderError(friendlyName, err)
+				}
+				field, err := someRecord.getField(key)
+				if err != nil {
+					return nil, newDecoderError(friendlyName, "Got unknown field %v", key)
+				}
+				fieldDatum, err := fieldCodecMap[field.Name].Decode(bytes.NewBuffer(b))
+				if err != nil {
+					return nil, newDecoderError(friendlyName, err)
+				}
+				field.Datum = fieldDatum
+			}
+			return someRecord, nil
+		},
+		ef: func(w io.Writer, datum interface{}) error {
+			// Record is Avro JSON encoded as a map with field names as key field values
+			// recursively Avro JSON encoded.
+
+			someRecord, ok := datum.(*Record)
+			if !ok {
+				return newEncoderError(friendlyName, "expected: Record; received: %T", datum)
+			}
+			if someRecord.Name != recordTemplate.Name {
+				return newEncoderError(friendlyName, "expected: %v; received: %v", recordTemplate.Name, someRecord.Name)
+			}
+
+			// Recursively Avro JSON encode each field in the right order.
+			var orderedMap OrderedMap
+			for idx, field := range someRecord.Fields {
+				var value interface{}
+				// check whether field datum is valid
+				if reflect.ValueOf(field.Datum).IsValid() {
+					value = field.Datum
+				} else if field.hasDefault {
+					value = field.defval
+				} else {
+					return newEncoderError(friendlyName, "field has no data and no default set: %v", field.Name)
+				}
+
+				// Avro encode each field value and then unmarshal back as we to finally stick
+				// it in a JSON map which gets marshalled out. Too many marshal and unmarshals!
+				var buff bytes.Buffer
+				tmpWriter := bufio.NewWriter(&buff)
```

Comment:
Why wrap the `*bytes.Buffer` here? A `*bytes.Buffer` is a `Reader` and `Writer` on its own